samcli/commands/package/s3_uploader.py

```diff
@@ -0,0 +1,210 @@
+            future.result()
+
+            return self.make_url(remote_path)
+
+        except botocore.exceptions.ClientError as ex:
+            error_code = ex.response["Error"]["Code"]
+            if error_code == "NoSuchBucket":
+                raise NoSuchBucketError(bucket_name=self.bucket_name)
+            raise ex
+
+    def upload_with_dedup(self, file_name, extension=None):
+        """
+        Makes and returns name of the S3 object based on the file's MD5 sum
+
+        :param file_name: file to upload
+        :param extension: String of file extension to append to the object
+        :return: S3 URL of the uploaded object
+        """
+
+        # This construction of remote_path is critical to preventing duplicate
+        # uploads of same object. Uploader will check if the file exists in S3
+        # and re-upload only if necessary. So the template points to same file
+        # in multiple places, this will upload only once
+
+        filemd5 = self.file_checksum(file_name)
+        remote_path = filemd5
+        if extension:
+            remote_path = remote_path + "." + extension
+
+        return self.upload(file_name, remote_path)
+
+    def file_exists(self, remote_path):
+        """
+        Check if the file we are trying to upload already exists in S3
+
+        :param remote_path:
+        :return: True, if file exists. False, otherwise
+        """
+
+        try:
+            # Find the object that matches this ETag
+            self.s3.head_object(Bucket=self.bucket_name, Key=remote_path)
+            return True
+        except botocore.exceptions.ClientError:
+            # Either File does not exist or we are unable to get
+            # this information.
+            return False
+
+    def make_url(self, obj_path):
+        return "s3://{0}/{1}".format(self.bucket_name, obj_path)
+
+    def file_checksum(self, file_name):
+
+        with open(file_name, "rb") as file_handle:
+            md5 = hashlib.md5()
+            # Read file in chunks of 4096 bytes
+            block_size = 4096
+
+            # Save current cursor position and reset cursor to start of file
+            curpos = file_handle.tell()
+            file_handle.seek(0)
+
+            buf = file_handle.read(block_size)
+            while buf:
+                md5.update(buf)
+                buf = file_handle.read(block_size)
+
+            # Restore file cursor's position
+            file_handle.seek(curpos)
+
+            return md5.hexdigest()
+
+    def to_path_style_s3_url(self, key, version=None):
+        """
+            This link describes the format of Path Style URLs
+            http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro
+        """
+        base = self.s3.meta.endpoint_url
+        result = "{0}/{1}/{2}".format(base, self.bucket_name, key)
+        if version:
+            result = "{0}?versionId={1}".format(result, version)
+
+        return result
+
+
+class ProgressPercentage:
```

Comment:
Should be be using what we do for Layers instead: https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/lib/utils/progressbar.py?

Reply:
I need to investigate if click's progress bar has a callback function. Click's own Progressbar class seems to be in a private py file `_term_ui_impl.py` and it also takes an iterable upfront to determine progress, which we dont know in this case. So, might need retro-fitting to use it.

Reply:
It is publicly exposed through `__init__.py`: https://github.com/pallets/click/blob/master/click/__init__.py#L36